### PR TITLE
High precision prices

### LINKED ISSUE #1756: High precision prices
Add support for high precision prices

### PR DESCRIPTION
**Pull request by acbeni**
# Description

Closes #1756

# Checklist

- [x] Update release Notes
- [x] Add to the current github milestone

### CODE REVIEW

File: commercetools-sdk-base/src/main/java/io/sphere/sdk/json/CentPrecisionMoneyRepresentation.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+import javax.money.MonetaryAmount;
+
+final class CentPrecisionMoneyRepresentation extends MoneyRepresentation{
+
+
+    @JsonCreator
+    public CentPrecisionMoneyRepresentation(Long centAmount, String currencyCode) {
```
Comment (katmatt) [inline commercetools-sdk-base/src/main/java/io/sphere/sdk/json/CentPrecisionMoneyRepresentation.java]: Please add `final`to the parameters.

File: commercetools-sdk-base/src/main/java/io/sphere/sdk/json/MoneyDeserializer.java [hunk 2]
```diff
@@ -19,6 +21,17 @@
     public MonetaryAmount deserialize(final JsonParser jsonParser, final DeserializationContext deserializationContext) throws IOException {
         final MoneyRepresentation moneyRepresentation = deserializationContext.readValue(jsonParser, MoneyRepresentation.class);
         final String currencyCode = moneyRepresentation.getCurrencyCode();
-        return MoneyImpl.ofCents(moneyRepresentation.getCentAmount(), currencyCode);
+        if(moneyRepresentation instanceof HighPrecisionMoneyRepresentation){
+            HighPrecisionMoneyRepresentation highPrecisionMoneyRepresentation = ((HighPrecisionMoneyRepresentation) moneyRepresentation);
```
Comment (katmatt) [inline commercetools-sdk-base/src/main/java/io/sphere/sdk/json/MoneyDeserializer.java]: Please add `final` here.

File: commercetools-sdk-base/src/main/java/io/sphere/sdk/json/MoneyDeserializer.java [hunk 3]
```diff
@@ -19,6 +21,17 @@
...
+            final int scale = highPrecisionMoneyRepresentation.getFractionDigits();
+            BigDecimal bigDecimal = highPrecisionMoneyRepresentation.getPreciseAmount().movePointLeft(scale);
```
Comment (katmatt) [inline commercetools-sdk-base/src/main/java/io/sphere/sdk/json/MoneyDeserializer.java]: Please add `final`here.

File: commercetools-sdk-base/src/main/java/io/sphere/sdk/json/MoneyRepresentation.java [hunk 4]
```diff
@@ -47,30 +52,39 @@ public String getCurrencyCode() {
...
         if(scale<0){
             scale = 0;
         }
-        number = number.setScale(scale, RoundingMode.DOWN);
-        return number.movePointRight(number.scale()).longValueExact();
+        number = number.setScale(scale, RoundingMode.HALF_EVEN);
+        return number.movePointRight(number.scale());
+    }
+
+    protected static BigDecimal queryFrom(MonetaryAmount amount, final int fractionsDigit) {
```
Comment (katmatt) [inline commercetools-sdk-base/src/main/java/io/sphere/sdk/json/MoneyRepresentation.java]: Please add `final` to the parameters.

File: commercetools-sdk-base/src/main/java/io/sphere/sdk/json/MoneyRepresentation.java [hunk 5]
```diff
@@ -47,30 +52,39 @@ public String getCurrencyCode() {
...
+                        .build());
         return monetaryAmount
                 .with(ROUNDING)
-                .query(MoneyRepresentation::queryFrom);
+                .query(amount -> queryFrom(monetaryAmount).longValue());
     }
 
-    private static Long queryFrom(MonetaryAmount amount) {
+
+    protected static BigDecimal queryFrom(MonetaryAmount amount) {
```
Comment (katmatt) [inline commercetools-sdk-base/src/main/java/io/sphere/sdk/json/MoneyRepresentation.java]: Please add `final` to the parameters.

File: commercetools-sdk-base/src/main/java/io/sphere/sdk/json/MoneyRepresentation.java [hunk 6]
```diff
@@ -14,29 +17,31 @@
 
 import static org.apache.commons.lang3.StringUtils.isEmpty;
 
-final class MoneyRepresentation {
+
+@JsonTypeInfo(
+        use = JsonTypeInfo.Id.NAME,
+        include = JsonTypeInfo.As.PROPERTY,
+        property = "type",
+        defaultImpl = CentPrecisionMoneyRepresentation.class
```
Comment (katmatt) [inline commercetools-sdk-base/src/main/java/io/sphere/sdk/json/MoneyRepresentation.java]: Nice that you defined this default/fallback here 🥇

File: commercetools-models/src/main/java/io/sphere/sdk/products/PriceDraftBuilder.java [hunk 7]
```diff
@@ -133,7 +133,7 @@ public PriceDraftDsl build() {
 
 
     public static PriceDraftBuilder of(final MonetaryAmount value) {
-        return new PriceDraftBuilder(MoneyImpl.of(value));
+        return new PriceDraftBuilder(value);
```
Comment (katmatt) [inline commercetools-models/src/main/java/io/sphere/sdk/products/PriceDraftBuilder.java]: Nice catch 🥇

**Comment by acbeni** [comment]
@katmatt code review fixes corrected 👍

**Pull request merged**